frappe/core/doctype/deleted_document/deleted_document.py

```diff
@@ -27,4 +36,27 @@ def restore(name):
 	deleted.restored = 1
 	deleted.db_update()
 
-	frappe.msgprint(_('Document Restored'))
+	if alert:
+		frappe.msgprint(_('Document Restored'))
+
+
+@frappe.whitelist()
+def bulk_restore(docnames):
+	docnames = frappe.parse_json(docnames)
+	message = _('Restoring Deleted Document')
+	restored = []
+
+	for i, d in enumerate(docnames):
+		try:
+			show_progress(docnames, message, i + 1, d)
+			restore(d, alert=False)
+			frappe.db.commit()
+			restored.append(d)
+
+		except frappe.DocumentAlreadyRestored:
+			pass
```

Comment:
The user will see multiple error messages in case a lot of already restored items are in the list. You can pop it from the message dict and show a bulk summary at once.